### PR TITLE
🚧 Ignore escaped characters

### PR DESCRIPTION
**Pull request by akaszynski**
This take a stab at ignoring escaped characters.  It's a basic (and probably inefficient) implementation, so if there's a better way to get this done, please let me know.

Also, the escape dictionary should be modifiable by the user.

### CODE REVIEW

**Comment by akaszynski** [comment]
Same problem here.  I've noticed that ``regex`` is almost always slower than the built-in ``replace``.

**Comment by akaszynski** [comment]
We're now using replace.  As for making the substitution dictionary customizable, would you prefer to have a file containing the key/value pairs, or a string that they read in?

**Comment by larsoner** [comment]
> As for making the substitution dictionary customizable, would you prefer to have a file containing the key/value pairs, or a string that they read in?

I find putting characters like `\n` in the command line to be annoying (usually takes me multiple tries to get it right), so I would lean toward a file

**Comment by akaszynski** [comment]
Agreed, I was encountering that issue as well when testing it out.  I'm thinking simple key value pairs in a csv, where the pairs are separated by line breaks.

```
"\n", " "
"\'", "'"
```

**Comment by larsoner** [comment]
Might make more sense to match the dictionary format

thing->replacement

**Comment by akaszynski** [comment]
As we’re going to have to use spaces, are quotes permitted in your dictionary format?
```
"\n"->” "
"\'"->"'"
```

**Comment by larsoner** [comment]
I would just make it that anything to the left of `->` is the substring to look for, and anything to the right is the replacement. So:
```
\n-> 
\'->
```

**Comment by akaszynski** [comment]
Should be good to review now.

File: codespell_lib/_codespell.py [hunk 1]
```diff
@@ -232,6 +232,10 @@ def parse_options(args):
                         help='Comma separated list of words to be ignored '
                              'by codespell. Words are case sensitive based on '
                              'how they are written in the dictionary file')
+    parser.add_argument('-P', '--sub-pairs', type=str, metavar='FILE',
+                        help='Custom substitution text file that contains '
+                             'substituions key value pairs.  Can be used to '
```
Comment (c72578) [inline codespell_lib/_codespell.py]: substituions->substitutions

File: codespell_lib/tests/test_basic.py [hunk 2]
```diff
@@ -73,6 +75,28 @@ def test_basic(tmpdir, capsys):
...
+def test_escaped(tmpdir, capsys):
+    """Test escaping characters"""
+    d = str(tmpdir)
+    with open(op.join(d, 'escaped_char.txt'), 'w') as f:
+        f.write(r"\n\nWe can")
+    assert cs.main(d) == 0
+
+
+def test_escaped_sub_file(tmpdir, capsys):
+    """Test escaping characters using substituion file"""
```
Comment (c72578) [inline codespell_lib/tests/test_basic.py]: substituion->substitution

File: codespell_lib/_codespell.py (hunk 1 above)
Comment (akaszynski) [reply codespell_lib/_codespell.py]: Funny, because this made it through the codespell check.

**Comment by larsoner** [comment]
Looks good. One last thing: we probably want to provide and use a default file, just like we do for the dictionary. Maybe it should just contain `\n-> ` to start.

**Comment by akaszynski** [comment]
> Looks good. One last thing: we probably want to provide and use a default file, just like we do for the dictionary. Maybe it should just contain `\n-> ` to start.

Perhaps `\'->'` as well.

**Comment by larsoner** [comment]
Fine with me

File: codespell_lib/_codespell.py [hunk 3]
```diff
@@ -232,6 +232,10 @@ def parse_options(args):
                         help='Comma separated list of words to be ignored '
                              'by codespell. Words are case sensitive based on '
                              'how they are written in the dictionary file')
+    parser.add_argument('-P', '--sub-pairs', type=str, metavar='FILE',
+                        help='Custom substitution text file that contains '
```
Comment (peternewman) [inline codespell_lib/_codespell.py]: I'm a bit unclear from the help detail what this is for, is it to "fix up" the dictionary to deal with it matching escape sequences? To actually do sed type runs on my codebase or something else?

Comment (peternewman) [reply codespell_lib/_codespell.py]: Is this linked to https://github.com/codespell-project/codespell/issues/233 ?

**Comment by larsoner** [comment]
Also in the related PR #174 it was noted that the `write-changes` flag was problematic when dealing with escapes, if we come back to this we should make sure that `write-changes` works even with these substitutions

**Comment by bl-ue** [comment]
Why the close?  I se this is quite old 🤔

**Comment by akaszynski** [comment]
> Why the close? I see this is quite old thinking

Just cleaning up old PRs.  I'd like to work on this, but this project hasn't been updated in quite some time (last release 6 months ago).

**Comment by peternewman** [comment]
Hi @akaszynski ,

Despite the lack of releases, coding is still happening (although mostly in the dictionary), also see #1923:
https://github.com/codespell-project/codespell/compare/v2.0.0...master

You didn't seem to have respond to the review comments myself and @larsoner had left if you were expecting it to have been merged.

**Comment by akaszynski** [comment]
> You didn't seem to have respond to the review comments myself and @larsoner had left if you were expecting it to have been merged.

True, there's still work to be done.  I'll work on this.

**Pull request closed without merging**